src/vs/workbench/contrib/notebook/browser/contrib/multicursor/notebookMulticursor.ts

```diff
@@ -0,0 +1,449 @@
+			));
+			controller.setSelections(new ViewModelEventsCollector(), undefined, match.selections, CursorChangeReason.Explicit);
+			this.cursorsControllers.push(controller);
+		});
+	}
+
+	private constructCoordinatesConverter(cell: ICellViewModel): ICoordinatesConverter {
+		return {
+			convertViewPositionToModelPosition(viewPosition: Position): Position {
+				return viewPosition;
+			},
+			convertViewRangeToModelRange(viewRange: Range): Range {
+				return viewRange;
+			},
+			validateViewPosition(viewPosition: Position, expectedModelPosition: Position): Position {
+				return viewPosition;
+			},
+			validateViewRange(viewRange: Range, expectedModelRange: Range): Range {
+				return viewRange;
+			},
+			convertModelPositionToViewPosition(modelPosition: Position, affinity?: PositionAffinity, allowZeroLineNumber?: boolean, belowHiddenRanges?: boolean): Position {
+				return modelPosition;
+			},
+			convertModelRangeToViewRange(modelRange: Range, affinity?: PositionAffinity): Range {
+				return modelRange;
+			},
+			modelPositionIsVisible(modelPosition: Position): boolean {
+				return true;
+			},
+			getModelLineViewLineCount(modelLineNumber: number): number {
+				return 1;
+			},
+			getViewLineNumberOfModelPosition(modelLineNumber: number, modelColumn: number): number {
+				return modelLineNumber;
+			}
+		};
+	}
+
+	private constructCursorSimpleModel(cell: ICellViewModel): ICursorSimpleModel {
+		return {
+			getLineCount(): number {
+				return cell.textBuffer.getLineCount();
+			},
+			getLineContent(lineNumber: number): string {
+				return cell.textBuffer.getLineContent(lineNumber);
+			},
+			getLineMinColumn(lineNumber: number): number {
+				return cell.textBuffer.getLineMinColumn(lineNumber);
+			},
+			getLineMaxColumn(lineNumber: number): number {
+				return cell.textBuffer.getLineMaxColumn(lineNumber);
+			},
+			getLineFirstNonWhitespaceColumn(lineNumber: number): number {
+				return cell.textBuffer.getLineFirstNonWhitespaceColumn(lineNumber);
+			},
+			getLineLastNonWhitespaceColumn(lineNumber: number): number {
+				return cell.textBuffer.getLineLastNonWhitespaceColumn(lineNumber);
+			},
+			normalizePosition(position: Position, affinity: PositionAffinity): Position {
+				return position;
+			},
+			getLineIndentColumn(lineNumber: number): number {
+				return indentOfLine(cell.textBuffer.getLineContent(lineNumber)) + 1;
+			}
+		};
+	}
+
+	private updateAnchorListeners() {
+		this.anchorDisposables.clear();
+
+		if (!this.anchorCell) {
+			throw new Error('Anchor cell is undefined');
+		}
+
+		// typing
+		this.anchorDisposables.add(this.anchorCell[1].onWillType((input) => {
+			this.state = NotebookMultiCursorState.Idle; // typing will continue to work as normal across ranges, just preps for another cmd+d
+			this.cursorsControllers.forEach(cursorController => {
+				cursorController.type(new ViewModelEventsCollector(), input, 'keyboard');
```

Comment:
after each `type`, "decorations" are updated properly for the cells in the viewport, but the ones outside of the viewport are still using "lazy" decorations, which are holding the old selection range, thus when we scroll them into the viewport, they are painted on the wrong range. We may want to dispose the decorations and update them.